tests/ui/self-span.stderr

```diff
@@ -1,27 +1,24 @@
-error[E0423]: expected value, found struct `S`
-  --> $DIR/self-span.rs:18:23
-   |
-3  | pub struct S {}
-   | --------------- `S` defined here
-...
-18 |         let _: Self = Self;
-   |                       ^^^^ help: use struct literal syntax instead: `S {}`
-
 error[E0308]: mismatched types
-  --> $DIR/self-span.rs:17:21
+  --> $DIR/self-span.rs:16:21
    |
+16 |     async fn method(self) {
+   |                     ^^^^ expected `()`, found struct `S`
 17 |         let _: () = self;
-   |                --   ^^^^ expected `()`, found struct `S`
-   |                |
-   |                expected due to this
+   |                -- expected due to this
+
+error: the `Self` constructor can only be used with tuple or unit structs
+  --> $DIR/self-span.rs:18:23
+   |
+18 |         let _: Self = Self;
+   |                       ^^^^ help: use curly brackets: `Self { /* fields */ }`
 
 error[E0308]: mismatched types
-  --> $DIR/self-span.rs:25:21
+  --> $DIR/self-span.rs:24:21
    |
+24 |     async fn method(self) {
+   |                     ^^^^ expected `()`, found enum `E`
 25 |         let _: () = self;
-   |                --   ^^^^ expected `()`, found enum `E`
-   |                |
-   |                expected due to this
+   |                -- expected due to this
```

Comment:
Do you have any insight into what's going on with this error message? The error before this PR is what I would expect and matches what you get in the non-trait case: https://play.rust-lang.org/?version=stable&mode=debug&edition=2018&gist=f8833ba4ae1fad764bbaa11f1f6f2701. The error after seems misplaced to me.

Reply:
Ah, interesting. I had read this message differently and believed it to be an improved version of the previous one, but I see now what the intention of the message is and how this is now a bit more confusing.

The span for that second `self`, which is renamed to `__self`, is set to the span of the receiver to fix #105. Besides being able to change only the resolution context, I'm not sure what heuristic to apply here to change only affected `self`s.

Reply:
I'm really not sure what to do here. The previous workaround seemed to "work" because we generated a new function. Now that we don't do so, we need to replace the span of all `self`s, not just of the one argument to the function as before. This will result in all error messages that correspond to `self` to point to the receiver, which is an unfortunate compromise.

My suggestion is that we gate this workaround to rustc versions <= 1.45, or whenever the fix landed in the compiler.

Reply:
Sounds good, let's do that. Thanks!

Reply:
Have a preference for library to use? My go-to is versioncheck.